mirascope/tools/_computer_use.py

```diff
@@ -0,0 +1,132 @@
+            cap_drop=["ALL"],  # Drop all capabilities
+        )
+
+    @classmethod
+    def _create_tar_stream(cls, files: dict[str, str]) -> io.BytesIO:
+        """Creates a tar stream from a dictionary of files."""
+        stream = io.BytesIO()
+        with tarfile.open(fileobj=stream, mode="w") as tar:
+            for name, content in files.items():
+                info = tarfile.TarInfo(name=name)
+                encoded_content = content.encode("utf-8")
+                info.size = len(encoded_content)
+                tar.addfile(info, io.BytesIO(encoded_content))
+        stream.seek(0)
+        return stream
+
+    @toolkit_tool
+    def execute_python(self, code: str, requirements: list[str] | None) -> str:
+        """Executes Python code in a Docker container.
+
+        docker_image: {self.__config__.docker_image}
+        allow_network: {self.__config__.allow_network}
+
+        Args:
+            self: ComputerUseToolkit instance
+            code: Python code to execute
+            requirements: List of Python package requirements to install
+        """
+
+        try:
+            contents = {
+                "main.py": code,
+            }
+            if requirements:
+                if not self._config().allow_network:
+                    return "Error: Network access is disabled. Cannot install requirements via pip."
+                contents["requirements.txt"] = "\n".join(requirements)
+            stream = self._create_tar_stream(contents)
+            with self._container_context() as container:
+                container.put_archive("/", stream)
+                if requirements:
+                    exit_code, output = container.exec_run(
+                        cmd=["pip", "install", "-r", "/requirements.txt"],
+                    )
+                    if exit_code != 0:
+                        return (
+                            f"Error installing requirements: {output.decode('utf-8')}"
+                        )
+                exit_code, output = container.exec_run(
+                    cmd=["python", "/main.py"],
+                )
+            return output.decode("utf-8")
+
+        except Exception as e:
+            return f"Error executing code: {str(e)}"
+
+    @toolkit_tool
+    def execute_shell(self, command: str) -> str:
+        """Executes shell commands in a Docker container.
+
+        docker_image: {self.__config__.docker_image}
+        allow_network: {self.__config__.allow_network}
+
+        Args:
+            command: Shell command to execute
+        """
+        try:
+            with self._container_context() as container:
+                exec_result = container.exec_run(
+                    cmd=["sh", "-c", command],
+                )
+            return exec_result.output.decode("utf-8")
+        except Exception as e:
+            return f"Error executing command: {str(e)}"
+
+    @contextmanager
```

Comment:
wait so if i'm understanding correctly, this will spin up a new container for every python/shell execution? It seems like instead the better thing to do would be to spin up a container at the beginning that the tool can then use? For example, what if the LLM wants to do something in one python call that relies on the outputs of a previous python call?